[export] add serialized_artifact test

ydwu4 opened the pull request:
Summary: This diff adds a EXPECTTEST_ACCEPT_ARTIFACT environment variable for test_serialize. When set to 1, it will save the ep to serialized_artifact and override existing saved artifact then load the saved artifact. When set to 0, it will just look up existing stored artifact and compare with the newly exported artifact.

Test Plan: test-only change.

Differential Revision: D72487556

angelayi commented on test/export/test_serialize.py:
```diff
@@ -49,6 +49,12 @@
 from torch.testing._internal.torchbind_impls import init_torchbind_implementations
 
 
+try:
```
should we only apply this checked-in artifact testing for a couple (< 10) artifacts? This seems like a lot of binaries being checked in.. not sure if pytorch is ok with this

ydwu4 replied:
not sure, let me talk to Richard

angelayi commented on test/export/test_serialize.py:
```diff
@@ -1466,6 +1671,7 @@ def forward(self, x):
 
 # We didn't set up kwargs input yet
 unittest.expectedFailure(TestDeserialize.test_exportdb_supported_case_fn_with_kwargs)
+unittest.expectedFailure(TestDeserialize.test_exportdb_supported_case_fn_with_kwargs)
```
delete?

angelayi commented on test/export/test_serialize.py:
```diff
@@ -859,6 +869,13 @@ def forward(self, x, y):
+            MyModule(),
+            inputs,
+            dynamic_shapes=dynamic_shapes,
+        )
```
how come we dont just directly apply this in `check_graph`?

ydwu4 committed:
[export] add serialized_artifact test (#152739)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/152739

This diff adds a EXPECTTEST_ACCEPT_ARTIFACT environment variable for test_serialize. When set to 1, it will save the ep to serialized_artifact and override existing saved artifact then load the saved artifact. When set to 0, it will just look up existing stored artifact and compare with the newly exported artifact.

Left some TODOs for test that cannot enable check_artifact yet due to graph module doesn't round trip: there are roughly three categories:
1. higher order op tuple output doesn't round trip
2. data dependent error
3. malicious node name difference causing graph to be different.

We omit meta data checks for now since the stack trace are not very stable when running in OSS vs in fbcode. If needed, we should also normalize them and check against the artifact.

Test Plan:
test-only change.

Rollback Plan:

Reviewed By: angelayi

Differential Revision: D72487556

ydwu4 commented on test/export/test_serialize.py:
```diff
@@ -59,6 +65,10 @@ def get_filtered_export_db_tests():
 class TestSerialize(TestCase):
+    def setUp(self):
+        # manually set rand seed for reproducibility
+        torch.manual_seed(0)
```
export will also save the params/buffers in state_dict in the binary. When load the binary back, we want to check the eager outputs of saved_ep.module() for saved artifact is the same as new_ep.module()'s output. If we don't set manual_seed(0) and when the model contains a randomly initialized param/buffer (e.g. test_module), their output don't match.
